File: SudokuCpp/SudokuSolver.cpp
#include <iostream>
#include <vector>
#include <chrono>

using namespace std;
using namespace std::chrono;

vector<vector<int>> board = {{0, 2, 0, 0, 0, 0, 0, 0, 0},
                             {0, 0, 0, 6, 0, 0, 0, 0, 3},
                             {0, 7, 4, 0, 8, 0, 0, 0, 0},
                             {0, 0, 0, 0, 0, 3, 0, 0, 2},
                             {0, 8, 0, 0, 4, 0, 0, 1, 0},
                             {6, 0, 0, 5, 0, 0, 0, 0, 0},
                             {0, 0, 0, 0, 1, 0, 7, 8, 0},
                             {5, 0, 0, 0, 0, 9, 0, 0, 0},
                             {0, 0, 0, 0, 0, 0, 0, 4, 0}};

int rowLen = board.size();
int colLen = board[0].size();
int gridHeight = 3;
int gridWidth = 3;
int numberChoices[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};

struct SquarePosition
{
    int rowPosition = -1, colPosition = -1;
};

void PrintSolveTime(high_resolution_clock::time_point startTime, high_resolution_clock::time_point endTime)
{
    auto timeTakenInSeconds = duration_cast<seconds>(endTime - startTime).count();
    auto timeTakenInMilliSeconds = duration_cast<milliseconds>(endTime - startTime).count();
    cout << "\nTime taken to solve : " << timeTakenInSeconds << "." << timeTakenInMilliSeconds % long(1E+3) << " seconds";
}

void PrintBoard()
{
    for (int row = 0; row < rowLen; ++row)
    {
        if (row % gridHeight == 0 && row != 0)
            cout << "------+-------+------" << endl;
        for (int col = 0; col < colLen; ++col)
        {
            if (col % gridWidth == 0 && col != 0)
                cout << "| ";
            if (col == colLen - 1)
                cout << board[row][col] << endl;
            else
                cout << board[row][col] << " ";
        }
    }
}

SquarePosition GetUnfilledPosition()
{
    SquarePosition square;
    for (int row = 0; row < rowLen; ++row)
    {
        for (int col = 0; col < colLen; ++col)
        {
            if (board[row][col] == 0)
            {
                square.rowPosition = row;
                square.colPosition = col;
                return square;
            }
        }
    }
    return square;
}

bool CanNumberBeInserted(int number, int currentRow, int currentCol)
{
    // check if number already exists in any other column in the same row
    for (int col = 0; col < colLen; ++col)
    {
        if (board[currentRow][col] == number && col != currentCol)
            return false;
    }
    // check if number already exists in any other row in the same column
    for (int row = 0; row < rowLen; ++row)
    {
        if (board[row][currentCol] == number && row != currentRow)
            return false;
    }
    // check if number already exists in any other position in the 3x3 grid
    int gridRowStartIndex = (currentRow / gridHeight) * 3, gridColStartIndex = (currentCol / gridWidth) * 3;
    int gridRowEndIndex = gridRowStartIndex + gridHeight, gridColEndIndex = gridColStartIndex + gridWidth;
    for (int row = gridRowStartIndex; row < gridRowEndIndex; ++row)
    {
        for (int col = gridColStartIndex; col < gridColEndIndex; ++col)
        {
            if (board[row][col] == number && row != currentRow && col != currentCol)
                return false;
        }
    }
    return true;
}

bool Solve()
{
    SquarePosition unfilledPosition = GetUnfilledPosition();
    if (unfilledPosition.rowPosition == -1 || unfilledPosition.colPosition == -1) // if no square is empty it means board is solved
        return true;
    int currentRow = unfilledPosition.rowPosition, currentCol = unfilledPosition.colPosition;
    for (int number : numberChoices)
    {
        if (CanNumberBeInserted(number, currentRow, currentCol))
        {
            board[currentRow][currentCol] = number;
            if (Solve())
                return true;
            // the board was unable to be solved at some point so reset all values till last solved position
            board[currentRow][currentCol] = 0;
        }
    }
    return false;
}

int main()
{
    cout << "\nBefore solving" << endl;
    PrintBoard();

    auto startTime = high_resolution_clock::now();
    bool isSolved = Solve();
    auto endTime = high_resolution_clock::now();

    if (isSolved)
    {
        cout << "\nAfter solving" << endl;
        PrintBoard();
    }
    else //unsolvable board
        cout << "The board is unsolvable" << endl;

    PrintSolveTime(startTime, endTime);

    return 0;
}

File: SudokuPython/SudokuSolver.py
import time

board = [
    [0, 2, 0, 0, 0, 0, 0, 0, 0],
    [0, 0, 0, 6, 0, 0, 0, 0, 3],
    [0, 7, 4, 0, 8, 0, 0, 0, 0],
    [0, 0, 0, 0, 0, 3, 0, 0, 2],
    [0, 8, 0, 0, 4, 0, 0, 1, 0],
    [6, 0, 0, 5, 0, 0, 0, 0, 0],
    [0, 0, 0, 0, 1, 0, 7, 8, 0],
    [5, 0, 0, 0, 0, 9, 0, 0, 0],
    [0, 0, 0, 0, 0, 0, 0, 4, 0]
]
row_len = len(board)
col_len = len(board[0])
grid_height = 3
grid_width = 3
number_choices = [1, 2, 3, 4, 5, 6, 7, 8, 9]


def getUnfilledSquare(board):
    for row in range(row_len):
        for col in range(col_len):
            if board[row][col] == 0:
                return row, col
    return None


def printBoard(board):
    for row in range(row_len):
        if row % grid_height == 0 and row != 0:
            print("------+-------+------")
        for col in range(col_len):
            if col % grid_width == 0 and col != 0:
                print("|", end=" ")
            if col == col_len-1:
                print(board[row][col])
            else:
                print(board[row][col], end=" ")


def canNumberBeInserted(board, number, current_position):
    current_row, current_col = current_position
    for col in range(col_len):
        if number == board[current_row][col] and col != current_col:
            return False
    for row in range(row_len):
        if number == board[row][current_col] and row != current_row:
            return False
    grid_row_start = (current_row//3)*3
    grid_col_start = (current_col//3)*3
    for row in range(grid_row_start, grid_row_start+grid_height):
        for col in range(grid_col_start, grid_col_start+grid_width):
            if number == board[row][col] and (row, col) != (current_row, current_col):
                return False
    return True


def solve(board):
    empty_square_position = getUnfilledSquare(board)
    if not empty_square_position:  # if no square is empty it means board is solved
        return True
    row, col = empty_square_position
    for number in number_choices:
        if canNumberBeInserted(board, number, (row, col)):
            board[row][col] = number
            if(solve(board)):
                return True
            # the board was unable to be solved at some point so reset all values till last solved
            board[row][col] = 0
    return False


def sudokuSolver():
    print("\nBefore solving\n")
    printBoard(board)
    start_time = time.time()
    solve(board)
    end_time = time.time()
    print("\n\nAfter solving\n")
    printBoard(board)
    print("\nTime taken to solve: {} seconds".format(end_time-start_time))


if __name__ == '__main__':
    sudokuSolver()
